### ISSUE TITLE
LiveStreamMonitor exception (unhandled internally)

### ISSUE THREAD

**Issue by Nectura** [TwitchLib/TwitchLib.Api#155]
Hello, the following happened after running the LiveStreamMonitor service for possibly a few days, im not entirely sure on the time. Surprisingly that somehow crashed my application.

Some usage context if that helps: I have a timer that runs every minute and calls "SetChannelsById"; I am only calling UpdateLiveStreamersAsync once, at the application start-up; there has been only 1 twitch channel in the list all along.

> 2020-10-22T00:21:55.639802057Z Unhandled exception. System.Net.Http.HttpRequestException: An error occurred while sending the request.
> 2020-10-22T00:21:55.639865762Z  ---> System.IO.IOException: Unable to read data from the transport connection: Connection reset by peer.
> 2020-10-22T00:21:55.639922366Z  ---> System.Net.Sockets.SocketException (104): Connection reset by peer
> 2020-10-22T00:21:55.639947168Z    --- End of inner exception stack trace ---
> 2020-10-22T00:21:55.639983171Z    at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
> 2020-10-22T00:21:55.639997172Z    at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.GetResult(Int16 token)
> 2020-10-22T00:21:55.640040775Z    at System.Net.Security.SslStream.<FillBufferAsync>g__InternalFillBufferAsync|215_0[TReadAdapter](TReadAdapter adap, ValueTask`1 task, Int32 min, Int32 initial)
> 2020-10-22T00:21:55.640055377Z    at System.Net.Security.SslStream.ReadAsyncInternal[TReadAdapter](TReadAdapter adapter, Memory`1 buffer)
> 2020-10-22T00:21:55.640088879Z    at System.Net.Http.HttpConnection.FillAsync()
> 2020-10-22T00:21:55.640102580Z    at System.Net.Http.HttpConnection.ReadNextResponseHeaderLineAsync(Boolean foldedHeadersAllowed)
> 2020-10-22T00:21:55.640144684Z    at System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage request, CancellationToken cancellationToken)
> 2020-10-22T00:21:55.640159085Z    --- End of inner exception stack trace ---
> 2020-10-22T00:21:55.640190987Z    at System.Net.Http.HttpConnection.SendAsyncCore(HttpRequestMessage request, CancellationToken cancellationToken)
> 2020-10-22T00:21:55.640204588Z    at System.Net.Http.HttpConnectionPool.SendWithNtConnectionAuthAsync(HttpConnection connection, HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
> 2020-10-22T00:21:55.656753594Z    at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
> 2020-10-22T00:21:55.656772195Z    at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
> 2020-10-22T00:21:55.656777496Z    at TwitchLib.Api.Core.Internal.TwitchHttpClientHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
> 2020-10-22T00:21:55.656781596Z    at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
> 2020-10-22T00:21:55.656787597Z    at TwitchLib.Api.Core.HttpCallHandlers.TwitchHttpClient.GeneralRequest(String url, String method, String payload, ApiVersion api, String clientId, String accessToken)
> 2020-10-22T00:21:55.656791797Z    at TwitchLib.Api.Core.ApiBase.<>c__DisplayClass14_0`1.<TwitchGetGenericAsync>b__1()
> 2020-10-22T00:21:55.656796097Z    at System.Threading.Tasks.Task`1.InnerInvoke()
> 2020-10-22T00:21:55.656800098Z    at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
> 2020-10-22T00:21:55.656812198Z --- End of stack trace from previous location where exception was thrown ---
> 2020-10-22T00:21:55.656815699Z    at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
> 2020-10-22T00:21:55.656830000Z --- End of stack trace from previous location where exception was thrown ---
> 2020-10-22T00:21:55.656833700Z    at TwitchLib.Api.Core.ApiBase.<>c__DisplayClass14_0`1.<<TwitchGetGenericAsync>b__0>d.MoveNext()
> 2020-10-22T00:21:55.656837801Z --- End of stack trace from previous location where exception was thrown ---
> 2020-10-22T00:21:55.656841101Z    at TwitchLib.Api.Services.LiveStreamMonitorService.GetLiveStreamersAsync()
> 2020-10-22T00:21:55.656844601Z    at TwitchLib.Api.Services.LiveStreamMonitorService.UpdateLiveStreamersAsync(Boolean callEvents)
> 2020-10-22T00:21:55.656848001Z    at TwitchLib.Api.Services.LiveStreamMonitorService.OnServiceTimerTick()
> 2020-10-22T00:21:55.656851402Z    at TwitchLib.Api.Services.Core.ServiceTimer.TimerElapsedAsync(Object sender, ElapsedEventArgs e)
> 2020-10-22T00:21:55.656854802Z    at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__139_1(Object state)
> 2020-10-22T00:21:55.656858502Z    at System.Threading.QueueUserWorkItemCallback.<>c.<.cctor>b__6_0(QueueUserWorkItemCallback quwi)
> 2020-10-22T00:21:55.656862202Z    at System.Threading.ExecutionContext.RunForThreadPoolUnsafe[TState](ExecutionContext executionContext, Action`1 callback, TState& state)
> 2020-10-22T00:21:55.656866003Z    at System.Threading.QueueUserWorkItemCallback.Execute()
> 2020-10-22T00:21:55.656869503Z    at System.Threading.ThreadPoolWorkQueue.Dispatch()

**Comment by sokiha** [comment]
This seems like a problem many people have had, see #113 and I can see you have also reported it [here.](https://github.com/TwitchLib/TwitchLib/issues/447) There has been made two PR to fix it. 

First #125  was created to fix it, but this was never merged. 
Instead #127  was created, and was merged into the dev branch. But this code change has magically disappeared from the dev branch and was never merged into the master branch.  At least I have not been able to find the merged change in any of these branches.

This error make the follower and LiveStreamerMonitor services practically useless as it will crash your program at random points in time eg. when a bad gateway error is received, without a good way to catch these exceptions, as they are raised on a thread from a thread pool, managed by the system. 

It would be nice if @swiftyspiffy could look into it, if he has time at some point, as he were the one who originally merged the fix in #127 to see what has happened with the change.

**Comment by sokiha** [comment]
I tried the fix from #127 on my own build of the Twitch.API. The change does not fix the problem just move where the exception is being thrown, as there is still no where it is catched by a try/catch. With the change the exception end being thrown on the following line in the modified version from #127 `Elapsed += async (sender, e) => await TimerElapsedAsync(sender, e);` https://github.com/TwitchLib/TwitchLib.Api/blob/a4de07db8f7a4bb516021aa9cd0ea5921e90f05a/TwitchLib.Api/Services/Core/ServiceTimer.cs#L19 
but as this is in an anonymous event handler method in the ServiceTimer constructor, you as a consumer of the LiveStreamMonitorService, is still not able to catch and handle the exception. For my own use I have just ended up adding a try/catch clause here https://github.com/TwitchLib/TwitchLib.Api/blob/a4de07db8f7a4bb516021aa9cd0ea5921e90f05a/TwitchLib.Api/Services/LiveStreamMonitorService.cs#L104-L108 Ignoring the http errors and missing an update is better than crashing the whole application, when your application should be able to run for a long time.
